File: algorithms/java/22_generate_parentheses/Generate.java
class Generate {
	public List<String> generateParentheses(int n) {
		List<String> res = new ArrayList<>();
		if (n <= 0) {
			return res;
		}
		generate(n, n, new StringBuilder(), res);
		return res;
	}
	private void generate(int left, int right, StringBuilder sb, List<String> res) {
		if (left == 0 && right == 0) {
            res.add(sb.toString());
            return;
        }
        if (left > 0) {
            sb.append('(');
            generate(left - 1, right, sb, res);
            sb.deleteCharAt(sb.length() - 1);
        }
        if (right > left) {
            sb.append(')');
            generate(left, right - 1, sb, res);
            sb.deleteCharAt(sb.length() - 1);
        }
    }
}

File: algorithms/cpp/22_generate_parentheses/Generate.cc
class Generate {
public:
	vector<string> generateParenthesis(int n) {
		vector<string> res;
		dfs(res, n, n, "");
		return res;
	}
private:
	void dfs(vector<string>& res, int left_remain, int right_remain, string current) {
		if (left == 0 && right == 0) {
			res.push_back(current);
			return;
		}
		if (left_remain > 0) {
			dfs(res, left_remain - 1, right_remain, current + "(");
		}
		if (right_remain > left_remain) {
			dfs(res, left_remain, right_remain - 1, current + ")");
		}
	}
};
